src/cargo/ops/cargo_output_metadata.rs

```diff
@@ -115,27 +115,46 @@ fn build_resolve_graph(
     let target_data = RustcTargetData::new(ws, &requested_kinds)?;
     // Resolve entire workspace.
     let specs = Packages::All.to_package_id_specs(ws)?;
+    let requested_features = RequestedFeatures::from_command_line(
+        &metadata_opts.features,
+        metadata_opts.all_features,
+        !metadata_opts.no_default_features,
+    );
     let resolve_opts = ResolveOpts::new(
         /*dev_deps*/ true,
         &metadata_opts.features,
         metadata_opts.all_features,
         !metadata_opts.no_default_features,
     );
+    let force_all = if metadata_opts.filter_platforms.is_empty() {
+        crate::core::resolver::features::ForceAllTargets::Yes
+    } else {
+        crate::core::resolver::features::ForceAllTargets::No
+    };
     let ws_resolve = ops::resolve_ws_with_opts(
         ws,
         &target_data,
         &requested_kinds,
         &resolve_opts,
         &specs,
         HasDevUnits::Yes,
-        crate::core::resolver::features::ForceAllTargets::No,
```

Comment:
This has be a little worried since it looks like it's changing default behavior. If you don't pass `--filter-platform` then this is switching from `No` to `Yes`? Off the top of my head I don't know the impact of that, however.

Reply:
Yeah, this one's a little weird. I think this actually means that metadata used to default to resolving for just the host target if no `--filter-platform` was specified, which doesn't really match what I'd expect from the name of the argument? But if that indeed used to be the default, I'm very happy to change this back! It does mean though that there is then no way to get `cargo metadata` to give info for all targets.

Reply:
Although looking at the test `metadata::filter_platform`, I feel like it should catch if the behavior actually changed here?

Reply:
I can try to explain the `::No` here.

When `ForceAllTargets` was added (#8269), it was only relevant to the new feature resolver (`-Zfeatures`).  Because `cargo metadata` does not use the new feature resolver, the value here wasn't really important (it was effectively ignored).

Afterwards, `download_accessible` was changed to look at it (#8823), I think I just neglected to update this line.  The consequence is that in some cases Cargo would perform downloads in two batches instead of one.

I think the change here seems correct.  I'll try to review this more deeply tomorrow, but from a high level it seems good.  It'll definitely be nice to avoid those extra downloads whenever running `cargo metadata` or `cargo tree`!

Reply:
Ok sounds good, thanks for weighing in @ehuss!